src/syntax/syntax.rkt

```diff
@@ -508,8 +509,35 @@
        (define vars (build-list (length ireprs) (lambda (i) (string->symbol (format "x~a" i)))))
        (synth-fl-impl name vars `(,new-op ,@vars))]))
 
+  ; make hash table
+  (define rules (make-hasheq))
+  (define commutes? #f)
+  (when identities
+    (print "looping")
+    (for ([ident (in-list identities)])
+      (match ident
+        [(list ident-name lhs-expr rhs-expr)
+         (cond
+           [(hash-has-key? rules ident-name)
+            (raise-herbie-syntax-error "Duplicate identity ~a" ident-name)]
+           [else (hash-set! rules ident-name (list lhs-expr rhs-expr))])]
+        [(list 'exact expr)
+         (hash-set! rules (gensym (string->symbol (format "~a-exact" name))) (list expr expr))]
```

Comment:
I'm a little worried about using `gensym`s here. I bet you could just number the rules.

Reply:
Also shouldn't it be, like, `(list expr (prog->spec expr))` or something?

Reply:
@bksaiki

Reply:
Agreed on both points. We can't call `prog->spec` here because of dependency cycles.